### PR TITLE
[SYCL][ESIMD]Add tests to validate lsc_block_load/store using 64 bit elements by default

### PR DESCRIPTION
**Pull request by fineg74**

### CODE REVIEW

**Comment by fineg74** [comment]
Complementary compiler PR: https://github.com/intel/llvm/pull/8316

**Comment by fineg74** [comment]
Test failures:
SYCL :: OneapiDeviceSelector/case_sensitivity.cpp
SYCL :: Reduction/reduction_ctor.cpp
Are not related to the change

File: SYCL/ESIMD/lsc/lsc_surf_load_u32.cpp [hunk 1]
```diff
@@ -32,6 +32,10 @@ template <int TestCastNum, typename T> bool tests() {
   passed &= test<TestCastNum + 8, T, 1, 4, 1, 32, true>();
   passed &= test<TestCastNum + 9, T, 2, 2, 1, 16, true>();
   passed &= test<TestCastNum + 10, T, 4, 4, 1, 4, true>();
+
+  // test lrge number of elements
```
Comment (sarnex) [inline SYCL/ESIMD/lsc/lsc_surf_load_u32.cpp]: nit: typo

**Review by sarnex — APPROVED** [review]

File: SYCL/ESIMD/lsc/Inputs/lsc_usm_block_load_prefetch.hpp [hunk 2]
```diff
@@ -143,6 +143,13 @@ template <typename T> bool test_lsc_block_load() {
     Passed &= test<T, 2, DS, L1H, L3H, NoPrefetch, CheckMerge>(Q, 5, 5);
   if constexpr (sizeof(T) >= sizeof(int))
     Passed &= test<T, 1, DS, L1H, L3H, NoPrefetch, CheckMerge>(Q, 3, 5);
+  // Only 512-bits maximum can be loaded at once (i.e. 4*128 bytes).
```
Comment (v-klochkov) [inline SYCL/ESIMD/lsc/Inputs/lsc_usm_block_load_prefetch.hpp:190]: @fineg74 - there was a conflicting change here that I resolved: https://github.com/intel/llvm-test-suite/pull/1590/commits/85f3753ae858be628f6380a6e908e4193443b6f9

But that seems to be not a complete fix. This test needs some additional adjustments such as passing overaligned<8> argument and aligned memory allocation

Comment (v-klochkov) [reply SYCL/ESIMD/lsc/Inputs/lsc_usm_block_load_prefetch.hpp:190]: Please add passing overaligned<8> tag for such cases.

Comment (v-klochkov) [reply SYCL/ESIMD/lsc/Inputs/lsc_usm_block_load_prefetch.hpp:190]: I couldn't find info in USM spec about the default alignment for malloc_shared().
Memory allocation on the line 42 _probably_ must use aligned_alloc_shared()

File: SYCL/ESIMD/lsc/lsc_usm_store_u32.cpp [hunk 3]
```diff
@@ -31,6 +31,11 @@ template <int TestCastNum, typename T> bool tests() {
   passed &= test<TestCastNum + 9, T, 2, 2, 1, 16, true>();
   passed &= test<TestCastNum + 10, T, 4, 4, 1, 4, true>();
 
+  // large number of elements
+  passed &= test<TestCastNum + 11, T, 4, 4, 1, 128, true,
```
Comment (v-klochkov) [inline SYCL/ESIMD/lsc/lsc_usm_store_u32.cpp:35]: 128 is correct only for T=int32_t, and is too much for T=int64_t.
Please use some checks, i.e. 'if constexpr (sizeof(T) <= 4)'

Can you please verify 256*int16_t and 512*int8_t ?

Please ensure that data is allocated with proper alignment.

File: SYCL/ESIMD/lsc/Inputs/lsc_usm_store.hpp [hunk 4]
```diff
@@ -82,7 +83,8 @@ bool test(uint32_t pmask = 0xffffffff) {
 
             if constexpr (transpose) {
               simd<T, VS> vals(new_val + elem_off, 1);
-              lsc_block_store<T, VS, DS, L1H, L3H>(out + elem_off, vals);
+              lsc_block_store<T, VS, DS, L1H, L3H>(out + elem_off, vals,
+                                                   Flags{});
```
Comment (v-klochkov) [inline SYCL/ESIMD/lsc/Inputs/lsc_usm_store.hpp]: Please verify both variants: with and without Flags parameter. For example:
```suggestion
              if constexpr (sizeof(T) < 8)
                  lsc_block_store<T, VS, DS, L1H, L3H>(out + elem_off, vals,
                                                       Flags{});
              else
                  lsc_block_store<T, VS, DS, L1H, L3H>(out + elem_off, vals);
```

File: SYCL/ESIMD/lsc/Inputs/lsc_surf_load.hpp [hunk 5]
```diff
@@ -17,11 +17,11 @@ using namespace sycl;
...
-template <int case_num, typename T, uint32_t Groups, uint32_t Threads,
-          uint16_t VL, uint16_t VS, bool transpose,
-          lsc_data_size DS = lsc_data_size::default_size,
-          cache_hint L1H = cache_hint::none, cache_hint L3H = cache_hint::none,
-          bool use_prefetch = false>
+template <
+    int case_num, typename T, uint32_t Groups, uint32_t Threads, uint16_t VL,
+    uint16_t VS, bool transpose, lsc_data_size DS = lsc_data_size::default_size,
+    cache_hint L1H = cache_hint::none, cache_hint L3H = cache_hint::none,
+    bool use_prefetch = false, typename Flags = __ESIMD_NS::overaligned_tag<4>>
```
Comment (v-klochkov) [inline SYCL/ESIMD/lsc/Inputs/lsc_surf_load.hpp:24]: The lines 64 and 65 below need to use std::vector constructor accepting alignment to guarantee it is at least 8 even for T=int8_t

File: SYCL/ESIMD/lsc/lsc_usm_store_u32.cpp (hunk 3 above)
Comment (fineg74) [reply SYCL/ESIMD/lsc/lsc_usm_store_u32.cpp:35]: This particular test verifies 32 bit data, there is a separate test for 64 bit data, I will add tests for 8 bit and 16 bit data.

File: SYCL/ESIMD/lsc/Inputs/lsc_usm_block_load_prefetch.hpp (hunk 2 above)
Comment (fineg74) [reply SYCL/ESIMD/lsc/Inputs/lsc_usm_block_load_prefetch.hpp:190]: As I see in the code malloc_shared(...) is translated into aligned_alloc_shared(alignof(T),....). The default alignment is 128

Comment (v-klochkov) [reply SYCL/ESIMD/lsc/Inputs/lsc_usm_block_load_prefetch.hpp:190]: Ok, thank you. That means that allocation of bytes guarantees only alignment by 1-byte. 
In practice that would be always a bigger alignment, but formally we have to use aligned_alloc_shared<int8_t>(8, ...) to be on safe side.

File: SYCL/ESIMD/lsc/lsc_usm_store_u8_u16.cpp [hunk 6]
```diff
@@ -25,6 +25,11 @@ template <int TestCastNum, typename T> bool tests() {
   passed &= test<TestCastNum + 4, T, 4, 4, 1, 256, true,
                  lsc_data_size::default_size, cache_hint::none,
                  cache_hint::none, __ESIMD_NS::overaligned_tag<8>>();
+  if constexpr (sizeof(T) == 1) {
```
Comment (v-klochkov) [inline SYCL/ESIMD/lsc/lsc_usm_store_u8_u16.cpp:28]: Thank you for adding this 'if constexpr (sizeof(T) == 1)' here. 
Shouldn't similar checks be added for the calls at the line 22 (sizeof(T) <= 4) and on the line 25 (sizeof(T) <= 2) ? 
How does that test pass when T=int32_t is passed and attempt to store 256 int32_t elements is done?

Comment (fineg74) [reply SYCL/ESIMD/lsc/lsc_usm_store_u8_u16.cpp:28]: This test tests 8 bit and 16bit data only, namely uint8_t and uint16_t, so int32_t is never tested here. 32 bit data is tested separately in tests that has u32 in their names (i.e. lsc_usm_store_u32, lsc_usm_store_u8u32, etc)

Comment (v-klochkov) [reply SYCL/ESIMD/lsc/lsc_usm_store_u8_u16.cpp:28]: Right, I missed the fact that this test is for bytes/shorts only. Thank you.

**Review by v-klochkov — APPROVED** [review]

**Pull request merged**